mpi4jax/_src/tokenizer.py

```diff
@@ -0,0 +1,154 @@
+    )
+    # Update bind_params to account for the additional token.
+    bind_params["num_carry"] += 1
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(token, *safe_map(read, eqn.invars), **bind_params)
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def while_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    new_body_fn = lambda token, *args: _token_forwarding(
+        jax.core.jaxpr_as_fun(bind_params["body_jaxpr"]), token
+    )(*args)
+    bind_params["body_jaxpr"] = jax.make_jaxpr(new_body_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # We use `auto_tokenize` here since the condition function
+    # is forced to only return a boolean value and cannot return
+    # a token.
+    new_cond_fn = lambda token, *args: auto_tokenize(
+        jax.core.jaxpr_as_fun(bind_params["cond_jaxpr"]), token
+    )(*args)
+    bind_params["cond_jaxpr"] = jax.make_jaxpr(new_cond_fn)(
+        token, *safe_map(read, eqn.invars)
+    )
+    # Update bind_params to account for the additional token.
+    ans = eqn.primitive.bind(token, *safe_map(read, eqn.invars), **bind_params)
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def cond_override(read, eqn, token):
+    _, bind_params = eqn.primitive.get_bind_params(eqn.params)
+    branch_jaxprs = []
+    cond_var = eqn.invars[0]
+    other_vars = eqn.invars[1:]
+    for branch in bind_params["branches"]:
+        new_branch_fn = lambda token, *args: _token_forwarding(
+            jax.core.jaxpr_as_fun(branch), token
+        )(*args)
+        new_jaxpr = jax.make_jaxpr(new_branch_fn)(token, *safe_map(read, other_vars))
+        branch_jaxprs.append(new_jaxpr)
+    bind_params["branches"] = tuple(branch_jaxprs)
+    bind_params["linear"] = (False,) + bind_params["linear"]
+    ans = eqn.primitive.bind(
+        read(cond_var), token, *safe_map(read, other_vars), **bind_params
+    )
+    token = ans[0]
+    ans = ans[1:]  # Drop the token.
+    return token, ans
+
+
+def _override_tokens(jaxpr, consts, token, *args):
+    if token is None:  # Create a new token if one is not passed.
+        token = jax.lax.create_token()
+
+    def read(v):
+        if type(v) is jax.core.Literal:
+            return v.val
+        else:
+            return env[v]
+
+    def write(v, val):
+        env[v] = val
+
+    env = {}
+    write(jax.core.unitvar, jax.core.unit)
+    safe_map(write, jaxpr.constvars, consts)
+    safe_map(write, jaxpr.invars, args)
+    for eqn in jaxpr.eqns:
+        # Here, we override the original primatives with our own
+        # token forcing method.
+        if eqn.primitive in token_override_registry:
+            token_override = token_override_registry[eqn.primitive]
+            ans = token_override(
+                safe_map(read, eqn.invars), new_token=token, **eqn.params
```

Comment:
What happens if `token` is given explicitly? E.g.

```python
@auto_tokenize
def foo(arr, token):
    return mpi4jax.allreduce(arr, op=MPI.SUM, token=token)
```

I think ideally we would emit a warning that the explicitly passed token is being dropped, but I'm not sure where to catch this case.

Reply:
IIUC the fundamental problem is that the `jaxpr` rewrite happens after all arguments are already bound to the corresponding primitives, so there's no way to know what was passed originally? In this case it *could* be solved with an extra boolean argument `token_passed` or so (that works similar to `transpose` for the `allreduce` primitive, in that it's only used internally and not exposed in the wrappers). But that's quite a price to pay in terms of code complexity just for raising a warning for something that doesn't actually break anything (passing tokens is just unnecessary). @PhilipVinc what do you think?

BTW, I like `auto_tokenize`.

Reply:
Ideally, I think it would be nice if we could expose a set of `mpi.auto.allreduce`, etc that do not return an extra token argument, and have the tokens be an internal detail that is handled by the decorator. It was @thenerdstation initial proposal and I liked it. That way you would have no way to pass tokens explicitly.

In C++, for example, you'd write operations in the correct order of execution.
I'm against adding complexity, instead documenting that operations in the `@auto_tokenize` scope will be executed in the order that you wrote them.

The only thing where care must be taken for this kind of things is in `vip/jvp` scopes, where the order of operations matter...

Reply:
OK, let's forget about the warning. We can decide how we want to simplify the API at a later stage, when we know the limitations of this approach.